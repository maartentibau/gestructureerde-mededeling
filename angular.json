{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "gestructureerde-mededeling": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:application",
          "options": {
            "allowedCommonJsDependencies": ["clipboard"],
            "outputPath": "dist/gestructureerde-mededeling",
            "index": "src/index.html",
            "browser": "src/main.ts",
            "tsConfig": "tsconfig.app.json",
            "inlineStyleLanguage": "scss",
            "assets": ["./src/favicon.ico", "./src/assets", "./src/_redirects"],
            "styles": [
              "normalize.css/normalize.css",
              "@angular/material/prebuilt-themes/azure-blue.css",
              "./src/styles.scss"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kB",
                  "maximumError": "1MB"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kB",
                  "maximumError": "4kB"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "gestructureerde-mededeling:build:production"
            },
            "development": {
              "buildTarget": "gestructureerde-mededeling:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n"
        },
        "test": {
          "builder": "@angular-devkit/build-angular:jest",
          "options": {
            "tsConfig": "tsconfig.spec.json"
          }
        }
      }
    }
  }
}
