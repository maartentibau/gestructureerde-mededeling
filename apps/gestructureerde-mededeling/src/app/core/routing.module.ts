import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CreateComponent } from '../components/create/create.component';
import { GenerateComponent } from '../components/generate/generate.component';
import { ValidateComponent } from '../components/validate/validate.component';

const routes: Routes = [
  { path: 'create', component: CreateComponent },
  { path: 'generate', component: GenerateComponent },
  { path: 'validate', component: ValidateComponent },
  { path: '', redirectTo: '/generate', pathMatch: 'full' },
  { path: '**', component: GenerateComponent },
];

@NgModule({
  imports: [RouterModule.forRoot(routes, { relativeLinkResolution: 'legacy' })],
  exports: [RouterModule],
})
export class RoutingModule {}
