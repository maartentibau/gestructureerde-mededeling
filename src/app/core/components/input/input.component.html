<div class="ogm-form">
  <input
    type="text"
    [attr.maxlength]="validate ? 12 : 10"
    [formControl]="ogmInput"
    [attr.placeholder]="placeholderMessage"
  />
</div>
