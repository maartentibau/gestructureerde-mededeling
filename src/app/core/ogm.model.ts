export type OgmData = {
  number: string | null;
  numberFormat: string | null;
};
