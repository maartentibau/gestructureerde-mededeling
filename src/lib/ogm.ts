export function ogm(): string {
  return 'ogm';
}
