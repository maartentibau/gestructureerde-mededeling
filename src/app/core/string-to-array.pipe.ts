import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  standalone: true,
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: number | string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
