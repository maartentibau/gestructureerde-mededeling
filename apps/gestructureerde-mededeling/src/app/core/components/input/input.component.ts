import { Component, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { FormControl } from '@angular/forms';
import { Observable, Subject } from 'rxjs';
import { filter, map, takeUntil, tap } from 'rxjs/operators';
import { OgmService } from '../../services/ogm.service';

export interface OgmInputChange {
  ogm: string;
  isValid: boolean;
}

@Component({
  selector: 'ogm-input',
  templateUrl: './input.component.html',
  styleUrls: ['./input.component.scss'],
})
export class InputComponent implements OnInit, OnDestroy {
  @Input() validate: boolean;
  @Input() placeholderMessage: string;

  @Output() ogmInputChange: EventEmitter<OgmInputChange> = new EventEmitter<undefined>();

  readonly ogmInput: FormControl;
  readonly destroy$: Subject<void>;

  ogm$: Observable<string>;

  constructor(private ogmService: OgmService) {
    this.ogmInput = new FormControl(null);
    this.destroy$ = new Subject<void>();
  }

  ngOnInit() {
    const ogmInitValue: string = this.ogmService.init();
    const hasSpaces: RegExp = new RegExp('[\\s]');

    const omgInputChanges$ = this.ogmInput.valueChanges.pipe(
      filter((value) => {
        if (isNaN(Number(value)) || hasSpaces.test(value)) {
          const cleanValue = this.ogmService.clean(value);
          this.ogmInput.setValue(cleanValue, { emitEvent: false });
          return false;
        }

        return true;
      })
    );

    if (this.validate) {
      this.ogm$ = omgInputChanges$.pipe(
        map((value) => (value ? this.ogmService.format(value.toString().padEnd(12, ' ')) : ogmInitValue))
      );
    } else {
      this.ogm$ = omgInputChanges$.pipe(map((value) => (value ? this.ogmService.generate(value, true) : ogmInitValue)));
    }

    this.ogm$
      .pipe(
        tap((ogm: string) => {
          let isValid;

          if (this.validate) {
            const ogmNumber = this.ogmService.clean(ogm);
            isValid = ogmNumber.length === 12 ? this.ogmService.validate(ogm) : undefined;
          }

          this.ogmInputChange.emit({ ogm, isValid });
        }),
        takeUntil(this.destroy$)
      )
      .subscribe();
  }

  ngOnDestroy() {
    this.destroy$.next();
  }
}
