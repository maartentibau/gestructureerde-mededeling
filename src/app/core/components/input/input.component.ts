import { ChangeDetectionStrategy, Component, inject, input, model } from '@angular/core';
import { OGM_EMPTY, OgmService } from '../../services/ogm.service';

export type Ogm = {
  ogm: string | null;
  isValid: boolean | null;
};

@Component({
  standalone: true,
  selector: 'ogm-input',
  templateUrl: './input.component.html',
  styleUrls: ['./input.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class InputComponent {
  readonly validate = input<boolean>(false);
  readonly placeholderMessage = input<string>('');

  readonly ogm = model<Ogm>({ ogm: null, isValid: null });

  #ogmService = inject(OgmService);

  handleInput(event: Event): void {
    let inputValue = (event.target as HTMLInputElement).value;
    const hasSpaces: RegExp = new RegExp('[\\s]');

    if (isNaN(Number(inputValue)) || hasSpaces.test(inputValue)) {
      (event.target as HTMLInputElement).value = this.#ogmService.clean(inputValue);
      return;
    }

    this.ogm.set(this.generateOgm(inputValue, this.validate()));
  }

  private generateOgm(value: string, validate: boolean): Ogm {
    let ogm = '';
    let isValid = null;

    if (validate) {
      ogm = value ? this.#ogmService.format(value.toString().padEnd(12, ' ')) : OGM_EMPTY;

      const ogmNumber = this.#ogmService.clean(ogm);
      isValid = ogmNumber.length === 12 ? this.#ogmService.validate(ogmNumber) : null;
    } else {
      ogm = value ? this.#ogmService.generate(value, true) : OGM_EMPTY;
    }

    return { ogm, isValid };
  }
}
