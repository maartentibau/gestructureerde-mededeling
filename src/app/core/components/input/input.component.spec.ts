import { TestBed } from '@angular/core/testing';
import { render, screen } from '@testing-library/angular';
import userEvent from '@testing-library/user-event';
import { OgmService } from '../../services/ogm.service';

import { InputComponent, Ogm } from './input.component';

describe('InputComponent', () => {
  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  it('should create', async () => {
    const { container } = await setup();

    expect(container).toBeTruthy();
  });

  describe('rendering component', () => {
    it('should match snapshot with no @Input properties defined', async () => {
      // prepare
      const { fixture } = await setup();

      // act
      await fixture.whenStable();

      // check
      expect(fixture).toMatchSnapshot();
    });

    it('should match snapshot with validate set to false', async () => {
      // prepare
      const { fixture } = await setup({ validate: false, placeholderMessage: 'Placeholder message' });

      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });

    it('should match snapshot with validate set to true', async () => {
      // prepare
      const { fixture } = await setup({ validate: true, placeholderMessage: 'Placeholder message' });

      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('handleInput', () => {
    let component: InputComponent;
    let ogmService: OgmService;
    let inputElement: HTMLElement;

    describe('without validation', () => {
      beforeEach(async () => {
        ({ component, ogmService } = await setup({ validate: false }));

        jest.spyOn(ogmService, 'generate');
        jest.spyOn(ogmService, 'clean');
        jest.spyOn(component.ogm, 'set');

        inputElement = screen.getByTestId('ogm-input');
      });

      describe('invalid characters', () => {
        it('should NOT set an ogm given invalid characters are used', async () => {
          // prepare
          const ogmInput: string = ' a b c,./;%$';

          // act
          await userEvent.type(inputElement, ogmInput);

          // check
          expect(ogmService.generate).not.toHaveBeenCalled();
          expect(ogmService.clean).toHaveBeenLastCalledWith('$');
          expect(component.ogm.set).not.toHaveBeenCalled();
        });
      });

      it('should emit a generated OGM and not run validation', async () => {
        // prepare
        const ogmInput: string = '12345';
        const expectedOgmInputChange: Ogm = { ogm: '+++123/4500/00012+++', isValid: null };

        jest.spyOn(ogmService, 'validate');

        // act
        await userEvent.type(inputElement, ogmInput);

        // check
        expect(ogmService.generate).toHaveBeenCalledWith(ogmInput, true);
        expect(ogmService.clean).not.toHaveBeenCalled();
        expect(ogmService.validate).not.toHaveBeenCalled();
        expect(component.ogm.set).toHaveBeenLastCalledWith(expectedOgmInputChange);
      });
    });

    describe('with validation', () => {
      beforeEach(async () => {
        ({ component, ogmService } = await setup({ validate: true }));

        jest.spyOn(ogmService, 'generate');
        jest.spyOn(ogmService, 'clean');
        jest.spyOn(ogmService, 'validate');
        jest.spyOn(component.ogm, 'set');

        inputElement = screen.getByTestId('ogm-input');
      });

      it('should emit a formatted OGM and NOT run validation when the input is not 12 characters long', async () => {
        // prepare
        const ogmInput: string = '12345';
        const ogmValue: string = '+++123/45  /     +++';
        const expectedOgmInputChange: Ogm = { ogm: ogmValue, isValid: null };

        // act
        await userEvent.type(inputElement, ogmInput);

        // check
        expect(ogmService.generate).not.toHaveBeenCalledWith();
        expect(ogmService.clean).toHaveBeenCalled();
        expect(ogmService.validate).not.toHaveBeenCalled();
        expect(component.ogm.set).toHaveBeenLastCalledWith(expectedOgmInputChange);
      });

      it('should emit a formatted OGM and run validation and mark the OGM as invalid', async () => {
        // prepare
        const ogmInput: string = '123456789012';
        const ogmValue: string = '+++123/4567/89012+++';
        const expectedOgmInputChange: Ogm = { ogm: ogmValue, isValid: false };

        // act
        await userEvent.type(inputElement, ogmInput);

        // check
        expect(ogmService.generate).not.toHaveBeenCalledWith();
        expect(ogmService.clean).toHaveBeenCalled();
        expect(ogmService.validate).toHaveBeenCalled();
        expect(component.ogm.set).toHaveBeenLastCalledWith(expectedOgmInputChange);
      });

      it('should emit a formatted OGM and run validation and mark the OGM as valid', async () => {
        // prepare
        const ogmInput: string = '165806877551';
        const ogmValue: string = '+++165/8068/77551+++';
        const expectedOgmInputChange: Ogm = { ogm: ogmValue, isValid: true };

        // act
        await userEvent.type(inputElement, ogmInput);

        // check
        expect(ogmService.generate).not.toHaveBeenCalledWith();
        expect(ogmService.clean).toHaveBeenCalled();
        expect(ogmService.validate).toHaveBeenCalledWith(ogmInput);
        expect(component.ogm.set).toHaveBeenLastCalledWith(expectedOgmInputChange);
      });
    });
  });
});

const setup = async (
  { validate, placeholderMessage }: { validate?: boolean; placeholderMessage?: string } = {
    validate: false,
    placeholderMessage: '',
  },
) => {
  const f = new OgmService();
  jest.spyOn(f, 'init');

  const renderResult = await render(InputComponent, {
    inputs: { validate, placeholderMessage },
    providers: [{ provide: OgmService, useValue: f }],
  });

  const ogmService = TestBed.inject(OgmService);

  return {
    ...renderResult,
    component: renderResult.fixture.componentInstance,
    ogmService,
  };
};
