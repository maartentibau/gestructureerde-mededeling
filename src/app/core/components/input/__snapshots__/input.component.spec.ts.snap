// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputComponent rendering component should match snapshot with no @Input properties defined 1`] = `
<ogm-input
  ogm={[Function Function]}
  placeholderMessage={[Function Function]}
  validate={[Function Function]}
>
  <div
    class="ogm-form"
  >
    <input
      autocomplete="off"
      data-tes=""
      data-testid="ogm-input"
      maxlength="10"
      placeholder=""
      type="text"
    />
  </div>
</ogm-input>
`;

exports[`InputComponent rendering component should match snapshot with validate set to false 1`] = `
<ogm-input
  ogm={[Function Function]}
  placeholderMessage={[Function Function]}
  validate={[Function Function]}
>
  <div
    class="ogm-form"
  >
    <input
      autocomplete="off"
      data-tes=""
      data-testid="ogm-input"
      maxlength="10"
      placeholder="Placeholder message"
      type="text"
    />
  </div>
</ogm-input>
`;

exports[`InputComponent rendering component should match snapshot with validate set to true 1`] = `
<ogm-input
  ogm={[Function Function]}
  placeholderMessage={[Function Function]}
  validate={[Function Function]}
>
  <div
    class="ogm-form"
  >
    <input
      autocomplete="off"
      data-tes=""
      data-testid="ogm-input"
      maxlength="12"
      placeholder="Placeholder message"
      type="text"
    />
  </div>
</ogm-input>
`;
